{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "\n",
    "library(BiocManager)\n",
    "library(BSgenome.Hsapiens.UCSC.hg38)\n",
    "library(ArchR)\n",
    "library(ggplot2)\n",
    "library(TFBSTools)\n",
    "library(Seurat)\n",
    "library(ggplot2)\n",
    "library(dplyr)\n",
    "library(harmony)\n",
    "#library(SeuratData)\n",
    "library(Signac)\n",
    "library(BSgenome.Hsapiens.UCSC.hg38)\n",
    "library(JASPAR2018)\n",
    "library(edgeR)\n",
    "\n",
    "data(\"geneAnnoHg38\")\n",
    "data(\"genomeAnnoHg38\")\n",
    "geneAnno <- geneAnnoHg38\n",
    "genomeAnno <- genomeAnnoHg38\n",
    "addArchRThreads(12)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#this load archr dev mode\n",
    "fn <- unclass(lsf.str(envir = asNamespace(\"ArchR\"), all = TRUE))\n",
    "  for(i in seq_along(fn)){\n",
    "    tryCatch({\n",
    "      eval(parse(text=paste0(fn[i], '<-ArchR:::', fn[i])))\n",
    "    }, error = function(x){\n",
    "    })\n",
    "  }"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "##Loading ArchR Projects as described in previous notebooks "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "mef_cleaned=read.table('getting the cleaned motifs as a bed file for Mef',sep='\\t',header=TRUE)\n",
    "names(mef_cleaned)<-c('seqnames','start','end','tfname')\n",
    "mef_cleaned$strand='*'\n",
    "#just a dummy score\n",
    "mef_cleaned$score=1.2\n",
    "head(mef_cleaned)\n",
    "mef_cleaned<-GRanges(mef_cleaned)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "#modfiying the getFootprints\n",
    "getFootprints<-function (ArchRProj = NULL, positions = NULL, plotName = \"Plot-Footprints\", \n",
    "    groupBy = \"Clusters\", useGroups = NULL, flank = 250, minCells = 25, \n",
    "    nTop = NULL, threads = getArchRThreads(), verbose = TRUE, \n",
    "    logFile = createLogFile(\"getFootprints\")) \n",
    "{\n",
    "    .validInput(input = ArchRProj, name = \"ArchRProj\", valid = c(\"ArchRProj\"))\n",
    "    .validInput(input = positions, name = \"positions\", valid = c(\"grangeslist\"))\n",
    "    .validInput(input = plotName, name = \"plotName\", valid = c(\"character\"))\n",
    "    .validInput(input = groupBy, name = \"groupBy\", valid = c(\"character\"))\n",
    "    .validInput(input = useGroups, name = \"useGroups\", valid = c(\"character\", \n",
    "        \"null\"))\n",
    "    .validInput(input = flank, name = \"flank\", valid = c(\"integer\"))\n",
    "    .validInput(input = minCells, name = \"minCells\", valid = c(\"integer\"))\n",
    "    .validInput(input = nTop, name = \"nTop\", valid = c(\"integer\", \n",
    "        \"null\"))\n",
    "    .validInput(input = threads, name = \"threads\", valid = c(\"integer\"))\n",
    "    .validInput(input = verbose, name = \"verbose\", valid = c(\"boolean\"))\n",
    "    .validInput(input = logFile, name = \"logFile\", valid = c(\"character\"))\n",
    "    tstart <- Sys.time()\n",
    "    .startLogging(logFile = logFile)\n",
    "    .logThis(mget(names(formals()), sys.frame(sys.nframe())), \n",
    "        \"Input-Parameters\", logFile = logFile)\n",
    "    coverageMetadata <- .getCoverageMetadata(ArchRProj = ArchRProj, \n",
    "        groupBy = groupBy, minCells = minCells)\n",
    "    coverageParams <- .getCoverageParams(ArchRProj = ArchRProj, \n",
    "        groupBy = groupBy)\n",
    "    kmerLength <- coverageParams$kmerLength\n",
    "    .logThis(coverageMetadata, \"coverageMetadata\", logFile = logFile)\n",
    "    .logThis(coverageParams, \"coverageParams\", logFile = logFile)\n",
    "    if (!is.null(useGroups)) {\n",
    "        if (sum(coverageMetadata[, 1] %in% useGroups) == 0) {\n",
    "            stop(\"No Groups found matching useGroups!\")\n",
    "        }\n",
    "        coverageMetadata <- coverageMetadata[coverageMetadata[, \n",
    "            1] %in% useGroups, ]\n",
    "    }\n",
    "    genome <- getGenome(ArchRProj)\n",
    "    .requirePackage(genome)\n",
    "    .requirePackage(\"Biostrings\", source = \"bioc\")\n",
    "    BSgenome <- eval(parse(text = genome))\n",
    "    BSgenome <- validBSgenome(BSgenome)\n",
    "    .logDiffTime(\"Computing Kmer Bias Table\", tstart, verbose = verbose, \n",
    "        logFile = logFile)\n",
    "    kmerTableList <- .kmerPositionFrequency(featureList = positions, \n",
    "        genome = BSgenome, flank = flank, k = kmerLength, threads = 1, \n",
    "        verbose = FALSE, logFile = logFile)\n",
    "    .logDiffTime(\"Computing Footprints\", tstart, verbose = verbose, \n",
    "        logFile = logFile)\n",
    "    footprintList <- .computeFootprints(featureList = positions, \n",
    "        coverageFiles = coverageMetadata$File, flank = flank, \n",
    "        threads = threads, verbose = FALSE, logFile = logFile)\n",
    "    .logDiffTime(\"Computing Footprints Bias\", tstart, verbose = verbose, \n",
    "        logFile = logFile)\n",
    "    footprintBiasList <- .computeFootprintsBias(kmerTableList = kmerTableList, \n",
    "        coverageFiles = coverageMetadata$File, threads = threads, \n",
    "        verbose = FALSE)\n",
    "    .logDiffTime(\"Summarizing Footprints\", tstart, verbose = verbose, \n",
    "        logFile = logFile)\n",
    "    footAssay <- lapply(seq_along(positions), function(x) {\n",
    "        footMat <- lapply(seq_along(footprintList), function(y) {\n",
    "            footprintList[[y]][, x]\n",
    "        }) %>% Reduce(\"cbind\", .)\n",
    "        colnames(footMat) <- coverageMetadata$Name\n",
    "        biasMat <- lapply(seq_along(footprintBiasList), function(y) {\n",
    "            footprintBiasList[[y]][, x]\n",
    "        }) %>% Reduce(\"cbind\", .)\n",
    "        colnames(biasMat) <- coverageMetadata$Name\n",
    "        rbind(footMat, biasMat)\n",
    "    }) %>% SimpleList\n",
    "    names(footAssay) <- names(positions)\n",
    "    rm(footprintList, footprintBiasList)\n",
    "    gc()\n",
    "    rowData <- DataFrame(x = c(seq(-flank, flank), seq(-flank, \n",
    "        flank)), type = c(rep(\"footprint\", flank * 2 + 1), rep(\"bias\", \n",
    "        flank * 2 + 1)))\n",
    "    se <- SummarizedExperiment::SummarizedExperiment(assays = footAssay, \n",
    "        colData = coverageMetadata, rowData = rowData)\n",
    "    metadata(se)$Params <- SimpleList(kmerLength = kmerLength, \n",
    "        flank = flank, date = Sys.Date())\n",
    "    return(se)\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "motifPositions <- getPositions(proj)\n",
    "motifs <- c(\"MEF2C\")\n",
    "markerMotifs <- unlist(lapply(motifs, function(x) grep(x, names(motifPositions), value = TRUE)))\n",
    "markerMotifs\n",
    "\n",
    "mef<-grep('MEF2C', names(motifPositions), value = TRUE)\n",
    "mef\n",
    "                              \n",
    "mef_cleaned=read.table('/oak/stanford/groups/akundaje/laks/illuminafiles/cardiomyogenesis/archr_input_output/aggregated_motif_files/VCM_GATA4_stringentcleaned.csv',\n",
    "                      sep='\\t',header=TRUE)\n",
    "names(mef_cleaned)<-c('seqnames','start','end','tfname')\n",
    "mef_cleaned$strand='*'\n",
    "#dummy score\n",
    "mef_cleaned$score=1.2\n",
    "head(mef_cleaned)\n",
    "mef_cleaned<-GRanges(mef_cleaned)\n",
    "\n",
    "                              \n",
    "                              \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "seFoot <- getFootprints(\n",
    "  ArchRProj = proj, \n",
    "  #if we need to the precleaned motifs, just uncomment the following line and comment the line after that \n",
    "  #positions = motifPositions[markerMotifs],\n",
    "  positions=GRangesList(mef_cleaned),\n",
    "  groupBy='Clusters1'\n",
    "    #useGroups=c('Ventricular_Cardiomyocytes')    \n",
    ")\n",
    "\n",
    "#renaming the se object. It wont work without giving it a name\n",
    "names(assays(seFoot))<-'MEF2C'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    ".groupMeans<-function (mat = NULL, groups = NULL, na.rm = TRUE, sparse = FALSE) \n",
    "{\n",
    "    #print('inside group means')\n",
    "    #print(groups)\n",
    "    stopifnot(!is.null(groups))\n",
    "    stopifnot(length(groups) == ncol(mat))\n",
    "    gm <- lapply(unique(groups), function(x) {\n",
    "        if (sparse) {\n",
    "            Matrix::rowMeans(mat[, which(groups == x), drop = F], \n",
    "                na.rm = na.rm)\n",
    "        }\n",
    "        else {\n",
    "            rowMeans(mat[, which(groups == x), drop = F], na.rm = na.rm)\n",
    "        }\n",
    "    }) %>% Reduce(\"cbind\", .)\n",
    "    #print('done with the group means')\n",
    "    #gm<-matrix(gm)\n",
    "    #print(unique(groups))\n",
    "    colnames(gm) <- unique(groups)\n",
    "    return(gm)\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    ".groupSds<-function (mat = NULL, groups = NULL, na.rm = TRUE, sparse = FALSE) \n",
    "{\n",
    "    stopifnot(!is.null(groups))\n",
    "    stopifnot(length(groups) == ncol(mat))\n",
    "    gs <- lapply(unique(groups), function(x) {\n",
    "        if (sparse) {\n",
    "            matrixStats::rowSds(as.matrix(mat[, which(groups == \n",
    "                x), drop = F]), na.rm = na.rm)\n",
    "        }\n",
    "        else {\n",
    "            matrixStats::rowSds(mat[, which(groups == x), drop = F], \n",
    "                na.rm = na.rm)\n",
    "        }\n",
    "    }) %>% Reduce(\"cbind\", .)\n",
    "    #gs<-matrix(gs)\n",
    "    colnames(gs) <- unique(groups)\n",
    "    return(gs)\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    ".ggFootprint<-function (seFoot = NULL, name = NULL, pal = NULL, smoothWindow = NULL, \n",
    "    flank = NULL, flankNorm = NULL, baseSize = 6, normMethod = NULL, \n",
    "    logFile = NULL) \n",
    "{\n",
    "    errorList <- list()\n",
    "    rowDF <- SummarizedExperiment::rowData(seFoot)\n",
    "    footMat <- .getAssay(seFoot[BiocGenerics::which(rowDF[, 2] == \n",
    "        \"footprint\"), ], name)\n",
    "    biasMat <- .getAssay(seFoot[BiocGenerics::which(rowDF[, 2] == \n",
    "        \"bias\"), ], name)\n",
    "    footDF <- rowDF[BiocGenerics::which(rowDF[, 2] == \"footprint\"), \n",
    "        ]\n",
    "    biasDF <- rowDF[BiocGenerics::which(rowDF[, 2] == \"bias\"), \n",
    "        ]\n",
    "    errorList$footMat <- footMat\n",
    "    errorList$biasMat <- biasMat\n",
    "    errorList$footDF <- footDF\n",
    "    errorList$biasDF <- biasDF\n",
    "    if (!is.null(smoothWindow)) {\n",
    "        .logMessage(\"Applying smoothing window to footprint\", \n",
    "            logFile = logFile)\n",
    "        footMat <- apply(footMat, 2, function(x) .centerRollMean(x, \n",
    "            smoothWindow))\n",
    "        biasMat <- apply(biasMat, 2, function(x) .centerRollMean(x, \n",
    "            smoothWindow))\n",
    "    }\n",
    "    .logMessage(\"Normalizing by flanking regions\", logFile = logFile)\n",
    "    idx <- which(abs(footDF$x) >= flank - flankNorm)\n",
    "    footMat <- t(t(footMat)/colMeans(footMat[idx, , drop = FALSE]))\n",
    "    biasMat <- t(t(biasMat)/colMeans(biasMat[idx, , drop = FALSE]))\n",
    "    errorList$footMatNorm <- footMat\n",
    "    errorList$biasMatNorm <- footMat\n",
    "    if (tolower(normMethod) == \"none\") {\n",
    "        title <- \"\"\n",
    "    }\n",
    "    else if (tolower(normMethod) == \"subtract\") {\n",
    "        title <- \"Tn5 Bias Subtracted\\n\"\n",
    "        footMat <- footMat - biasMat\n",
    "        #print('Here in ggfootprint Subtract')\n",
    "    }\n",
    "    else if (tolower(normMethod) == \"divide\") {\n",
    "        title <- \"Tn5 Bias Divided\\n\"\n",
    "        footMat <- footMat/biasMat\n",
    "    }\n",
    "    else {\n",
    "        stop(\"normMethod not recognized!\")\n",
    "    }\n",
    "    .logMessage(paste0(\"NormMethod = \", normMethod), logFile = logFile)\n",
    "    #print('Here in ggfootprint block done ')\n",
    "    print(SummarizedExperiment::colData(seFoot)$Group)\n",
    "    print(dim(footMat))\n",
    "    footMatMean <- .groupMeans(footMat, SummarizedExperiment::colData(seFoot)$Group)\n",
    "    #footMatMean <- .groupMeans(footMat, 'Ventricular_Cardiomyocytes')\n",
    "    footMatSd <- .groupSds(footMat, SummarizedExperiment::colData(seFoot)$Group)\n",
    "    biasMatMean <- .groupMeans(biasMat, SummarizedExperiment::colData(seFoot)$Group)\n",
    "    biasMatSd <- .groupSds(biasMat, SummarizedExperiment::colData(seFoot)$Group)\n",
    "    smoothFoot <- rowMaxs(apply(footMatMean, 2, function(x) .centerRollMean(x, \n",
    "        11)))\n",
    "    errorList$footMatMean <- footMatMean\n",
    "    errorList$footMatSd <- footMatSd\n",
    "    errorList$biasMatMean <- biasMatMean\n",
    "    errorList$biasMatSd <- biasMatSd\n",
    "    errorList$smoothFoot <- smoothFoot\n",
    "    plotIdx <- seq_len(nrow(footMatMean))\n",
    "    plotFootDF <- lapply(seq_len(ncol(footMatMean)), function(x) {\n",
    "        data.frame(x = footDF$x, mean = footMatMean[, x], sd = footMatSd[, \n",
    "            x], group = colnames(footMatMean)[x])[plotIdx, , \n",
    "            drop = FALSE]\n",
    "    }) %>% Reduce(\"rbind\", .)\n",
    "    plotFootDF$group <- factor(paste0(plotFootDF$group), levels = unique(gtools::mixedsort(paste0(plotFootDF$group))))\n",
    "    plotBiasDF <- lapply(seq_len(ncol(biasMatMean)), function(x) {\n",
    "        data.frame(x = biasDF$x, mean = biasMatMean[, x], sd = biasMatSd[, \n",
    "            x], group = colnames(biasMatMean)[x])[plotIdx, , \n",
    "            drop = FALSE]\n",
    "    }) %>% Reduce(\"rbind\", .)\n",
    "    plotBiasDF$group <- factor(paste0(plotBiasDF$group), levels = unique(gtools::mixedsort(paste0(plotBiasDF$group))))\n",
    "    errorList$plotFootDF <- plotFootDF\n",
    "    errorList$plotBiasDF <- plotBiasDF\n",
    "    out <- tryCatch({\n",
    "        if (is.null(pal)) {\n",
    "            pal <- paletteDiscrete(values = gtools::mixedsort(SummarizedExperiment::colData(seFoot)$Group))\n",
    "        }\n",
    "        plotMax <- plotFootDF[order(plotFootDF$mean, decreasing = TRUE), \n",
    "            ]\n",
    "        plotMax <- plotMax[abs(plotMax$x) > 20 & abs(plotMax$x) < \n",
    "            50, ]\n",
    "        plotMax <- plotMax[!duplicated(plotMax$group), ]\n",
    "        plotMax <- plotMax[seq_len(ceiling(nrow(plotMax)/4)), \n",
    "            ]\n",
    "        plotMax$x <- 25\n",
    "        ggFoot <- ggplot(plotFootDF, aes(x = x, y = mean, color = group)) + \n",
    "            geom_ribbon(aes(ymin = mean - sd, ymax = mean + sd, \n",
    "                linetype = NA, fill = group), alpha = 0.4) + \n",
    "            geom_line() + scale_color_manual(values = pal) + \n",
    "            scale_fill_manual(values = pal) + xlab(\"Distance to motif center (bp)\") + \n",
    "            coord_cartesian(expand = FALSE, ylim = c(quantile(plotFootDF$mean, \n",
    "                1e-04), 1.15 * quantile(smoothFoot, 0.999)), \n",
    "                xlim = c(min(plotFootDF$x), max(plotFootDF$x))) + \n",
    "            theme_ArchR(baseSize = baseSize) + ggtitle(name) + \n",
    "            guides(fill = FALSE) + guides(color = FALSE) + ylab(paste0(title, \n",
    "            \"Normalized Insertions\")) + ggrepel::geom_label_repel(data = plotMax, \n",
    "            aes(label = group), size = 3, xlim = c(75, NA))\n",
    "        ggBias <- ggplot(plotBiasDF, aes(x = x, y = mean, color = group)) + \n",
    "            geom_ribbon(aes(ymin = mean - sd, ymax = mean + sd, \n",
    "                linetype = NA, fill = group), alpha = 0.4) + \n",
    "            geom_line() + scale_color_manual(values = pal) + \n",
    "            scale_fill_manual(values = pal) + xlab(\"Distance to motif center (bp)\") + \n",
    "            coord_cartesian(expand = FALSE, ylim = c(quantile(plotBiasDF$mean, \n",
    "                1e-04), 1.05 * quantile(plotBiasDF$mean, 0.999)), \n",
    "                xlim = c(min(plotBiasDF$x), max(plotBiasDF$x))) + \n",
    "            theme_ArchR(baseSize = baseSize) + ylab(\"Tn5-Bias Normalized Insertions\") + \n",
    "            theme(legend.position = \"bottom\", legend.box.background = element_rect(color = NA))\n",
    "        ggAlignPlots(ggFoot, .ggSmallLegend(ggBias), sizes = c(2, \n",
    "            1), draw = FALSE)\n",
    "    }, error = function(e) {\n",
    "        .logError(e, fn = \".ggFootprint\", info = name, errorList = errorList, \n",
    "            logFile = logFile)\n",
    "    })\n",
    "    out\n",
    "}\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "plotFootprints<-function (seFoot = NULL, names = NULL, pal = NULL, flank = 250, \n",
    "    flankNorm = 50, normMethod = \"Subtract\", smoothWindow = NULL, \n",
    "    baseSize = 6, plot = TRUE, ArchRProj = NULL, plotName = paste0(\"Plot-Footprints-\", \n",
    "        normMethod), height = 6, width = 4, addDOC = TRUE, force = FALSE, \n",
    "    logFile = createLogFile(\"plotFootprints\")) \n",
    "{\n",
    "    .validInput(input = seFoot, name = \"seFoot\", valid = c(\"SummarizedExperiment\"))\n",
    "    .validInput(input = names, name = \"names\", valid = c(\"character\", \n",
    "        \"null\"))\n",
    "    .validInput(input = pal, name = \"pal\", valid = c(\"palette\", \n",
    "        \"null\"))\n",
    "    .validInput(input = flank, name = \"flank\", valid = c(\"integer\"))\n",
    "    .validInput(input = flankNorm, name = \"flankNorm\", valid = c(\"integer\"))\n",
    "    .validInput(input = normMethod, name = \"normMethod\", valid = c(\"character\"))\n",
    "    .validInput(input = smoothWindow, name = \"smoothWindow\", \n",
    "        valid = c(\"integer\", \"null\"))\n",
    "    .validInput(input = baseSize, name = \"baseSize\", valid = c(\"numeric\"))\n",
    "    .validInput(input = plot, name = \"plot\", valid = c(\"boolean\"))\n",
    "    .validInput(input = ArchRProj, name = \"ArchRProj\", valid = c(\"ArchRProj\"))\n",
    "    .validInput(input = plotName, name = \"plotName\", valid = c(\"character\"))\n",
    "    .validInput(input = height, name = \"height\", valid = c(\"numeric\"))\n",
    "    .validInput(input = width, name = \"width\", valid = c(\"numeric\"))\n",
    "    .validInput(input = addDOC, name = \"addDOC\", valid = c(\"boolean\"))\n",
    "    .validInput(input = force, name = \"force\", valid = c(\"boolean\"))\n",
    "    .validInput(input = logFile, name = \"logFile\", valid = c(\"character\"))\n",
    "    tstart <- Sys.time()\n",
    "    .startLogging(logFile = logFile)\n",
    "    if (is.null(names)) {\n",
    "        names <- names(assays(seFoot))\n",
    "    }\n",
    "    if (length(names) > 25) {\n",
    "        if (!plot) {\n",
    "            if (force) {\n",
    "                .logMessage(\"Plotting more than 25 footprints can create large storage of ggplots. Continuing since force = TRUE\", \n",
    "                  verbose = TRUE, logFile = logFile)\n",
    "            }\n",
    "            else {\n",
    "                .logStop(\"Plotting more than 25 footprints can create large storage of ggplots. Stopping since force = FALSE\", \n",
    "                  logFile = logFile)\n",
    "            }\n",
    "        }\n",
    "    }\n",
    "    if (plot) {\n",
    "        name <- gsub(\"\\\\.pdf\", \"\", plotName)\n",
    "        if (is.null(ArchRProj)) {\n",
    "            outDir <- \"Plots\"\n",
    "        }\n",
    "        else {\n",
    "            ArchRProj <- .validArchRProject(ArchRProj)\n",
    "            outDir <- file.path(getOutputDirectory(ArchRProj), \n",
    "                \"Plots\")\n",
    "        }\n",
    "        dir.create(outDir, showWarnings = FALSE)\n",
    "        if (addDOC) {\n",
    "            doc <- gsub(\":\", \"-\", stringr::str_split(Sys.time(), \n",
    "                pattern = \" \", simplify = TRUE)[1, 2])\n",
    "            filename <- file.path(outDir, paste0(name, \"_Date-\", \n",
    "                Sys.Date(), \"_Time-\", doc, \".pdf\"))\n",
    "        }\n",
    "        else {\n",
    "            filename <- file.path(outDir, paste0(name, \".pdf\"))\n",
    "        }\n",
    "        pdf(filename, width = width, height = height, useDingbats = FALSE)\n",
    "    }\n",
    "    ggList <- lapply(seq_along(names), function(x) {\n",
    "        print(x)\n",
    "        print(names[1])\n",
    "        .logDiffTime(sprintf(\"Plotting Footprint : %s (%s of %s)\", \n",
    "            names[x], x, length(names)), t1 = tstart, logFile = logFile, \n",
    "            verbose = TRUE)\n",
    "        gg <- .ggFootprint(seFoot = seFoot, name = names[x], \n",
    "            pal = pal, smoothWindow = smoothWindow, flank = flank, \n",
    "            flankNorm = flankNorm, baseSize = baseSize, normMethod = normMethod, \n",
    "            logFile = logFile)\n",
    "        print(gg)\n",
    "        if (plot) {\n",
    "            if (x != 1) {\n",
    "                grid::grid.newpage()\n",
    "            }\n",
    "            grid::grid.draw(gg)\n",
    "            return(0)\n",
    "        }\n",
    "        else {\n",
    "            return(gg)\n",
    "        }\n",
    "    })\n",
    "    .endLogging(logFile = logFile)\n",
    "    if (!plot) {\n",
    "        names(ggList) <- names\n",
    "        ggList\n",
    "    }\n",
    "    else {\n",
    "        dev.off()\n",
    "        return(invisible(0))\n",
    "    }\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "plotFootprints(\n",
    "  seFoot = seFoot,\n",
    "  ArchRProj = proj, \n",
    "  normMethod = \"Divide\",\n",
    "  plotName = \"MEF2C-Footprints-stringentcleaned\",\n",
    "  addDOC = FALSE\n",
    "  ,\n",
    "    smoothWindow = 15,\n",
    "    plot=TRUE,\n",
    "    pal=c(\n",
    "\"Atrial_Cardiomyocytes\"=\"#FFAC53\",\"Ventricular_Cardiomyocytes\"=\"#F15F30\",\"Myocardium\"=\"#FFE500\"\n",
    ",\"Cardiac_Fibroblast\"=\"#3B9AB2\",\"Cardiac_fibroblast_progenitors\"=\"#0AD7D3\",\"vSMC\"=\"#A8DDB5\",\"Pericytes\"=\"#79FFFF\"\n",
    ",\"Neural_Crest\"=\"#D0CD47\",\"Early_Cardiac_fibroblast\"=\"#90D5E4\",\"Early_OFT_SMC\"=\"#AAD962\",\"Undifferentiated_epicardium\"=\"#208A42\"\n",
    ",\"Endocardial_cushion\"=\"#74A9FF\",\"Endocardial_cushion_late\"=\"#74A9CF\",\"Vasculatur_development\"=\"#74C476\",\"Endocaridum\"=\"#C06CAB\",\"Endocaridum_transitioning\"=\"#C06CFF\",\"Venal_endothelium\"=\"#EC6BB1\"\n",
    ",\"Endocaridum_unidentifed\"=\"#89288F\",\"Capillary_endothelium\"=\"#E6C2DC\",\"Arterial_endothelium\"=\"#F37B7D\"\n",
    ")\n",
    ")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "R",
   "language": "R",
   "name": "ir"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "4.0.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
