{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-08-26T09:20:55.736111Z",
     "start_time": "2020-08-26T09:20:55.440793Z"
    }
   },
   "outputs": [],
   "source": [
    "import math\n",
    "import tensorflow\n",
    "from tensorflow.compat.v1.keras.backend import get_session\n",
    "tensorflow.compat.v1.disable_v2_behavior()\n",
    "import kerasAC \n",
    "from kerasAC.generators.snp_generator import *\n",
    "from kerasAC.tiledb_config import *\n",
    "from scipy.special import softmax\n",
    "from kerasAC.interpret.deepshap import * \n",
    "from kerasAC.interpret.profile_shap import * \n",
    "from kerasAC.vis import * \n",
    "from kerasAC.helpers.transform_bpnet_io import * "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "chd_pcgc=pd.read_csv('chd_final_0base.csv',sep='\\t')\n",
    "controls=pd.read_csv('chd_control_final_0base.csv',sep='\\t')\n",
    "\n",
    "print(controls.shape)\n",
    "print(chd_pcgc.shape)\n",
    "#controls.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "dict_var_Score_cases={}\n",
    "dict_var_Score_controls={}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "import glob\n",
    "import re\n",
    "#from keras.models import load_model, model_from_json\n",
    "#from keras.utils.generic_utils import get_custom_objects\n",
    "from tensorflow.keras.models import load_model, model_from_json\n",
    "from tensorflow.keras.utils import get_custom_objects\n",
    "from kerasAC.metrics import * \n",
    "from kerasAC.custom_losses import * \n",
    "from kerasAC.metrics import * \n",
    "from kerasAC.custom_losses import * \n",
    "#from keras_genomics.layers.convolutional import RevCompConv1D\n",
    "import tensorflow_probability as tfp\n",
    "peak_files=glob.glob('mutations/chd_cases_*_ext.bed')\n",
    "print(peak_files)\n",
    "print(len(peak_files))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "for zz in peak_files:\n",
    "    name=re.sub('_$','',zz.split('ext')[0].split('chd_cases_')[1].replace('all_','all'))\n",
    "    print(name)\n",
    "    case_snps_in_peaks=pd.read_csv('mutations/chd_cases_'+str(name)+'_ext.bed',sep='\\t',header=None)\n",
    "    case_snps_in_peaks.columns=['Chrom','Pos0','Pos']\n",
    "    case_snps_in_peaks.shape\n",
    "    control_snps_in_peaks=pd.read_csv('mutations/chd_controls_'+str(name)+'_ext.bed',sep='\\t',header=None)\n",
    "    control_snps_in_peaks.columns=['Chrom','Pos0','Pos']\n",
    "    snps_inpeaks=chd_pcgc.merge(case_snps_in_peaks,on=['Chrom','Pos0','Pos'])\n",
    "    snps_inpeaks=snps_inpeaks.drop_duplicates()\n",
    "    print(snps_inpeaks.shape)\n",
    "    control_snps_inpeaks=controls.merge(control_snps_in_peaks,on=['Chrom','Pos0','Pos'])\n",
    "    control_snps_inpeaks=control_snps_inpeaks.drop_duplicates()\n",
    "    print(control_snps_inpeaks.shape)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#scoring mutations\n",
    "for zz in peak_files[13:14]:\n",
    "    name=re.sub('_$','',zz.split('ext')[0].split('chd_cases_')[1].replace('all_','all'))\n",
    "    print(name)\n",
    "    case_snps_in_peaks=pd.read_csv('mutations/chd_cases_'+str(name)+'_ext.bed',sep='\\t',header=None)\n",
    "    case_snps_in_peaks.columns=['Chrom','Pos0','Pos']\n",
    "    case_snps_in_peaks.shape\n",
    "    control_snps_in_peaks=pd.read_csv('mutations/chd_controls_'+str(name)+'_ext.bed',sep='\\t',header=None)\n",
    "    control_snps_in_peaks.columns=['Chrom','Pos0','Pos']\n",
    "    snps_inpeaks=chd_pcgc.merge(case_snps_in_peaks,on=['Chrom','Pos0','Pos'])\n",
    "    print(snps_inpeaks.shape)\n",
    "    snps_inpeaks=snps_inpeaks.drop_duplicates()\n",
    "    control_snps_inpeaks=controls.merge(control_snps_in_peaks,on=['Chrom','Pos0','Pos'])\n",
    "    control_snps_inpeaks=control_snps_inpeaks.drop_duplicates()\n",
    "    print(control_snps_inpeaks.shape)\n",
    "    snps_inpeaks.to_csv(\"chd_final_0base_\"+str(name)+\".tsv\",header=True,index=False,sep='\\t')\n",
    "    control_snps_inpeaks.to_csv(\"chd_control_final_0base_\"+str(name)+\".tsv\",header=True,index=False,sep='\\t')\n",
    "    for zz1 in range(0,5):\n",
    "        #load the model! \n",
    "        model_num=str(zz1)\n",
    "        custom_objects={\"recall\":recall,\n",
    "                    \"sensitivity\":recall,\n",
    "                    \"specificity\":specificity,\n",
    "                    \"fpr\":fpr,\n",
    "                    \"fnr\":fnr,\n",
    "                    \"precision\":precision,\n",
    "                    \"f1\":f1,\n",
    "                    \"ambig_binary_crossentropy\":ambig_binary_crossentropy,\n",
    "                    \"ambig_mean_absolute_error\":ambig_mean_absolute_error,\n",
    "                    \"ambig_mean_squared_error\":ambig_mean_squared_error,\n",
    "                    \"MultichannelMultinomialNLL\":MultichannelMultinomialNLL}\n",
    "        get_custom_objects().update(custom_objects)\n",
    "        import json \n",
    "        model=model_from_json(open(\"weights/\"+name+\".\"+str(model_num)+\".arch\",'r').read())\n",
    "        model.load_weights(\"weights/\"+name+\".\"+str(model_num)+\".weights\")\n",
    "        #reference allele sequence generator \n",
    "        ref_gen=SNPGenerator(bed_path=\"chd_final_0base_\"+name+\".tsv\",\n",
    "                 chrom_col=\"Chrom\",\n",
    "                 pos_col=\"Pos0\",\n",
    "                 allele_col=\"Ref\",\n",
    "                 rsid_col='rsid',\n",
    "                 flank_size=1057,\n",
    "                 ref_fasta=\"GRCh38.fasta\",\n",
    "                 expand_dims=False)\n",
    "        #alternate allele sequence generator \n",
    "        alt_gen=SNPGenerator(bed_path=\"chd_final_0base_\"+name+\".tsv\",\n",
    "                 chrom_col=\"Chrom\",\n",
    "                 pos_col=\"Pos0\",\n",
    "                 allele_col=\"Alt\",\n",
    "                 rsid_col='rsid',\n",
    "                 flank_size=1057,\n",
    "                 ref_fasta=\"GRCh38.fasta\",\n",
    "                 expand_dims=False)\n",
    "        #get the reference allele predictions \n",
    "        count_preds={} \n",
    "        profile_preds={} \n",
    "        snp_to_seq={} \n",
    "        for i in range(0,len(ref_gen)):\n",
    "            print(str(i))\n",
    "            cur_x=ref_gen[i] \n",
    "            batch_rsids=cur_x[0] \n",
    "            batch_preds=model.predict(cur_x[1])\n",
    "            batch_preds_profile=batch_preds[0]\n",
    "            batch_preds_count=batch_preds[1] \n",
    "            for batch_index in range(len(batch_rsids)): \n",
    "                cur_rsid=batch_rsids[batch_index]\n",
    "                snp_to_seq[cur_rsid]={} \n",
    "                snp_to_seq[cur_rsid]['ref']=cur_x[1][batch_index,:,:]\n",
    "                cur_pred_profile=batch_preds_profile[batch_index,:,:]\n",
    "                cur_pred_count=batch_preds_count[batch_index,:][0]\n",
    "                count_preds[cur_rsid]={}\n",
    "                count_preds[cur_rsid]['ref']=cur_pred_count\n",
    "                profile_preds[cur_rsid]={}\n",
    "                profile_preds[cur_rsid]['ref']=cur_pred_profile \n",
    "        #get the alternate allele predictions \n",
    "        for i in range(0,len(alt_gen)):\n",
    "            print(str(i))\n",
    "            cur_x=alt_gen[i] \n",
    "            batch_rsids=cur_x[0] \n",
    "            batch_preds=model.predict(cur_x[1])\n",
    "            batch_preds_profile=batch_preds[0]\n",
    "            batch_preds_count=batch_preds[1] \n",
    "            for batch_index in range(len(batch_rsids)): \n",
    "                cur_rsid=batch_rsids[batch_index]\n",
    "                snp_to_seq[cur_rsid]['alt']=cur_x[1][batch_index,:,:]\n",
    "                cur_pred_profile=batch_preds_profile[batch_index,:,:]\n",
    "                cur_pred_count=batch_preds_count[batch_index,:][0]\n",
    "                count_preds[cur_rsid]['alt']=cur_pred_count\n",
    "                profile_preds[cur_rsid]['alt']=cur_pred_profile \n",
    "        count_preds_cases=count_preds\n",
    "        profile_preds_cases=profile_preds\n",
    "        snp_to_seq_cases=snp_to_seq\n",
    "        #\n",
    "        #sum and then log\n",
    "        log_sumprob_ratios_100=[]\n",
    "        log_sumprob_ratios_100[:]=[]\n",
    "        for i in profile_preds_cases.keys():\n",
    "            pred_prob_ref=softmax(profile_preds_cases[i]['ref'],axis=0)\n",
    "            pred_prob_alt=softmax(profile_preds_cases[i]['alt'],axis=0)\n",
    "            log_sumprob_ratios_100.append(np.log(np.sum(pred_prob_ref[450:550])/np.sum(pred_prob_alt[450:550])))\n",
    "        #\n",
    "        local_sumcount_changes_100=[]\n",
    "        local_sumcount_changes_100[:]=[]\n",
    "        for i in profile_preds_cases.keys():\n",
    "            pred_prob_ref=softmax(profile_preds_cases[i]['ref'],axis=0)\n",
    "            pred_prob_alt=softmax(profile_preds_cases[i]['alt'],axis=0)\n",
    "            count_pred_ref=count_preds_cases[i]['ref']\n",
    "            count_pred_alt=count_preds_cases[i]['alt']\n",
    "            local_sumcount_changes_100.append(np.log( ((np.sum(np.exp(count_pred_ref)*pred_prob_ref[450:550])) / (np.sum(np.exp(count_pred_alt)*pred_prob_alt[450:550])))))\n",
    "        #\n",
    "        local_count_changes_100=[]\n",
    "        local_count_changes_100[:]=[]\n",
    "        for i in profile_preds_cases.keys():\n",
    "            pred_prob_ref=softmax(profile_preds_cases[i]['ref'],axis=0)\n",
    "            pred_prob_alt=softmax(profile_preds_cases[i]['alt'],axis=0)\n",
    "            count_pred_ref=count_preds_cases[i]['ref']\n",
    "            count_pred_alt=count_preds_cases[i]['alt']\n",
    "            local_count_changes_100.append(np.sum(np.exp(count_pred_ref)*pred_prob_ref[450:550]-np.exp(count_pred_alt)*pred_prob_alt[450:550]))\n",
    "        profile_preds_df=pd.DataFrame.from_dict(profile_preds_cases,orient='index')\n",
    "        profile_preds_df['local_sumcount_changes_100']=local_sumcount_changes_100\n",
    "        #sum and log \n",
    "        profile_preds_df['log_sumprob_ratios_100']=log_sumprob_ratios_100\n",
    "        #reference allele sequence generator \n",
    "        ref_gen=SNPGenerator(bed_path='chd_control_final_0base_'+str(name)+'.tsv',\n",
    "                 chrom_col=\"Chrom\",\n",
    "                 pos_col=\"Pos0\",\n",
    "                 allele_col=\"Ref\",\n",
    "                 rsid_col='rsid',\n",
    "                 flank_size=1057,\n",
    "                 ref_fasta=\"GRCh38.fasta\",\n",
    "                 expand_dims=False)\n",
    "        #alternate allele sequence generator \n",
    "        alt_gen=SNPGenerator(bed_path='chd_control_final_0base_'+str(name)+'.tsv',\n",
    "                 chrom_col=\"Chrom\",\n",
    "                 pos_col=\"Pos0\",\n",
    "                 allele_col=\"Alt\",\n",
    "                 rsid_col='rsid',\n",
    "                 flank_size=1057,\n",
    "                 ref_fasta=\"GRCh38.fasta\",\n",
    "                 expand_dims=False)\n",
    "        #get the reference allele predictions \n",
    "        count_preds={} \n",
    "        profile_preds={} \n",
    "        snp_to_seq={} \n",
    "        for i in range(0,len(ref_gen)):\n",
    "            print(str(i))\n",
    "            cur_x=ref_gen[i] \n",
    "            batch_rsids=cur_x[0] \n",
    "            batch_preds=model.predict(cur_x[1])\n",
    "            batch_preds_profile=batch_preds[0]\n",
    "            batch_preds_count=batch_preds[1] \n",
    "            for batch_index in range(len(batch_rsids)): \n",
    "                cur_rsid=batch_rsids[batch_index]\n",
    "                snp_to_seq[cur_rsid]={} \n",
    "                snp_to_seq[cur_rsid]['ref']=cur_x[1][batch_index,:,:]\n",
    "                cur_pred_profile=batch_preds_profile[batch_index,:,:]\n",
    "                cur_pred_count=batch_preds_count[batch_index,:][0]\n",
    "                count_preds[cur_rsid]={}\n",
    "                count_preds[cur_rsid]['ref']=cur_pred_count\n",
    "                profile_preds[cur_rsid]={}\n",
    "                profile_preds[cur_rsid]['ref']=cur_pred_profile \n",
    "        #get the alternate allele predictions \n",
    "        for i in range(0,len(alt_gen)):\n",
    "            print(str(i))\n",
    "            cur_x=alt_gen[i] \n",
    "            batch_rsids=cur_x[0] \n",
    "            batch_preds=model.predict(cur_x[1])\n",
    "            batch_preds_profile=batch_preds[0]\n",
    "            batch_preds_count=batch_preds[1] \n",
    "            for batch_index in range(len(batch_rsids)): \n",
    "                cur_rsid=batch_rsids[batch_index]\n",
    "                snp_to_seq[cur_rsid]['alt']=cur_x[1][batch_index,:,:]\n",
    "                cur_pred_profile=batch_preds_profile[batch_index,:,:]\n",
    "                cur_pred_count=batch_preds_count[batch_index,:][0]\n",
    "                count_preds[cur_rsid]['alt']=cur_pred_count\n",
    "                profile_preds[cur_rsid]['alt']=cur_pred_profile \n",
    "        count_preds_controls=count_preds\n",
    "        profile_preds_controls=profile_preds\n",
    "        snp_to_seq_controls=snp_to_seq\n",
    "        #sum and then log\n",
    "        log_sumprob_ratios_100=[]\n",
    "        #\n",
    "        log_sumprob_ratios_100[:]=[]\n",
    "        for i in profile_preds_controls.keys():\n",
    "            pred_prob_ref=softmax(profile_preds_controls[i]['ref'],axis=0)\n",
    "            pred_prob_alt=softmax(profile_preds_controls[i]['alt'],axis=0)\n",
    "            log_sumprob_ratios_100.append(np.log(np.sum(pred_prob_ref[450:550])/np.sum(pred_prob_alt[450:550])))\n",
    "        #\n",
    "        local_sumcount_changes_100=[]\n",
    "        local_sumcount_changes_100[:]=[]\n",
    "        for i in profile_preds_controls.keys():\n",
    "            pred_prob_ref=softmax(profile_preds_controls[i]['ref'],axis=0)\n",
    "            pred_prob_alt=softmax(profile_preds_controls[i]['alt'],axis=0)\n",
    "            count_pred_ref=count_preds_controls[i]['ref']\n",
    "            count_pred_alt=count_preds_controls[i]['alt']\n",
    "            local_sumcount_changes_100.append(np.log( ((np.sum(np.exp(count_pred_ref)*pred_prob_ref[450:550])) / (np.sum(np.exp(count_pred_alt)*pred_prob_alt[450:550])))))\n",
    "        #\n",
    "        profile_preds_df_controls=pd.DataFrame.from_dict(profile_preds_controls,orient='index')\n",
    "        #sum and log counts\n",
    "        profile_preds_df_controls['local_sumcount_changes_100']=local_sumcount_changes_100\n",
    "        #sum and log \n",
    "        profile_preds_df_controls['log_sumprob_ratios_100']=log_sumprob_ratios_100\n",
    "        dict_var_Score_cases[name+'_'+str(model_num)]=profile_preds_df\n",
    "        dict_var_Score_controls[name+'_'+str(model_num)]=profile_preds_df_controls\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "dict_var_Score_cases.keys()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "dict_var_Score_sumprob_cases_100_odds={}\n",
    "dict_var_Score_sumprob_controls_100_odds={}\n",
    "#\n",
    "dict_var_Score_sumcounts_cases_100={}\n",
    "dict_var_Score_sumcounts_controls_100={}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import tensorflow_probability as tfp\n",
    "peak_files=glob.glob('mutations/chd_cases_*_ext.bed')\n",
    "for zz in peak_files[13:14]:\n",
    "    name=re.sub('_$','',zz.split('ext')[0].split('chd_cases_')[1].replace('all_','all'))\n",
    "    print(name)\n",
    "    c0_fold0_cases=dict_var_Score_cases[name+'_0']\n",
    "    c0_fold0_controls=dict_var_Score_controls[name+'_0']\n",
    "    c0_fold1_cases=dict_var_Score_cases[name+'_1']\n",
    "    c0_fold1_controls=dict_var_Score_controls[name+'_1']\n",
    "    c0_fold2_cases=dict_var_Score_cases[name+'_2']\n",
    "    c0_fold2_controls=dict_var_Score_controls[name+'_2']\n",
    "    c0_fold3_cases=dict_var_Score_cases[name+'_3']\n",
    "    c0_fold3_controls=dict_var_Score_controls[name+'_3']\n",
    "    c0_fold4_cases=dict_var_Score_cases[name+'_4']\n",
    "    c0_fold4_controls=dict_var_Score_controls[name+'_4']\n",
    "    del c0_fold0_cases['ref']\n",
    "    del c0_fold0_cases['alt']\n",
    "    del c0_fold0_controls['ref']\n",
    "    del c0_fold0_controls['alt']\n",
    "    #\n",
    "    del c0_fold1_cases['ref']\n",
    "    del c0_fold1_cases['alt']\n",
    "    del c0_fold1_controls['ref']\n",
    "    del c0_fold1_controls['alt']\n",
    "    #\n",
    "    del c0_fold2_cases['ref']\n",
    "    del c0_fold2_cases['alt']\n",
    "    del c0_fold2_controls['ref']\n",
    "    del c0_fold2_controls['alt']\n",
    "    #\n",
    "    del c0_fold3_cases['ref']\n",
    "    del c0_fold3_cases['alt']\n",
    "    del c0_fold3_controls['ref']\n",
    "    del c0_fold3_controls['alt']\n",
    "    #\n",
    "    del c0_fold4_cases['ref']\n",
    "    del c0_fold4_cases['alt']\n",
    "    del c0_fold4_controls['ref']\n",
    "    del c0_fold4_controls['alt']\n",
    "    #\n",
    "    c0_fold0_cases.columns=['local_sumcount_changes_100_1','log_sumprob_ratios_100_1']\n",
    "    c0_fold1_cases.columns=['local_sumcount_changes_100_2','log_sumprob_ratios_100_2']\n",
    "    c0_fold2_cases.columns=['local_sumcount_changes_100_3','log_sumprob_ratios_100_3']\n",
    "    c0_fold3_cases.columns=['local_sumcount_changes_100_4','log_sumprob_ratios_100_4']\n",
    "    c0_fold4_cases.columns=['local_sumcount_changes_100_5','log_sumprob_ratios_100_5']\n",
    "    c0_fold0_controls.columns=['local_sumcount_changes_100_1','log_sumprob_ratios_100_1']\n",
    "    c0_fold1_controls.columns=['local_sumcount_changes_100_2','log_sumprob_ratios_100_2']\n",
    "    c0_fold2_controls.columns=['local_sumcount_changes_100_3','log_sumprob_ratios_100_3']\n",
    "    c0_fold3_controls.columns=['local_sumcount_changes_100_4','log_sumprob_ratios_100_4']\n",
    "    c0_fold4_controls.columns=['local_sumcount_changes_100_5','log_sumprob_ratios_100_5']\n",
    "    #\n",
    "    #log of sum counts\n",
    "    #\n",
    "    c0_local_sumcount_changes_100_cases=c0_fold0_cases[['local_sumcount_changes_100_1']].merge(c0_fold1_cases[['local_sumcount_changes_100_2']],left_index=True,right_index=True).\\\n",
    "    merge(c0_fold2_cases[['local_sumcount_changes_100_3']],left_index=True,right_index=True).\\\n",
    "    merge(c0_fold3_cases[['local_sumcount_changes_100_4']],left_index=True,right_index=True).\\\n",
    "    merge(c0_fold4_cases[['local_sumcount_changes_100_5']],left_index=True,right_index=True)\n",
    "    #\n",
    "    c0_local_sumcount_changes_100_controls=c0_fold0_controls[['local_sumcount_changes_100_1']].merge(c0_fold1_controls[['local_sumcount_changes_100_2']],left_index=True,right_index=True).\\\n",
    "    merge(c0_fold2_controls[['local_sumcount_changes_100_3']],left_index=True,right_index=True).\\\n",
    "    merge(c0_fold3_controls[['local_sumcount_changes_100_4']],left_index=True,right_index=True).\\\n",
    "    merge(c0_fold4_controls[['local_sumcount_changes_100_5']],left_index=True,right_index=True)\n",
    "    ###############\n",
    "    #log of sum\n",
    "    #\n",
    "    c0_local_sumprob_ratios_odd100_cases=c0_fold0_cases[['log_sumprob_ratios_100_1']].merge(c0_fold1_cases[['log_sumprob_ratios_100_2']],left_index=True,right_index=True).\\\n",
    "    merge(c0_fold2_cases[['log_sumprob_ratios_100_3']],left_index=True,right_index=True).\\\n",
    "    merge(c0_fold3_cases[['log_sumprob_ratios_100_4']],left_index=True,right_index=True).\\\n",
    "    merge(c0_fold4_cases[['log_sumprob_ratios_100_5']],left_index=True,right_index=True)\n",
    "    #\n",
    "    c0_local_sumprob_ratios_odd100_controls=c0_fold0_controls[['log_sumprob_ratios_100_1']].merge(c0_fold1_controls[['log_sumprob_ratios_100_2']],left_index=True,right_index=True).\\\n",
    "    merge(c0_fold2_controls[['log_sumprob_ratios_100_3']],left_index=True,right_index=True).\\\n",
    "    merge(c0_fold3_controls[['log_sumprob_ratios_100_4']],left_index=True,right_index=True).\\\n",
    "    merge(c0_fold4_controls[['log_sumprob_ratios_100_5']],left_index=True,right_index=True)\n",
    "    #\n",
    "    ######sumcounts\n",
    "    c0_local_sumcount_changes_100_cases['mean']=np.mean(c0_local_sumcount_changes_100_cases,axis=1)\n",
    "    c0_local_sumcount_changes_100_controls['mean']=np.mean(c0_local_sumcount_changes_100_controls,axis=1)\n",
    "    c0_local_sumcount_changes_200_cases['mean']=np.mean(c0_local_sumcount_changes_200_cases,axis=1)\n",
    "    c0_local_sumcount_changes_200_controls['mean']=np.mean(c0_local_sumcount_changes_200_controls,axis=1)\n",
    "    c0_local_sumcount_changes_300_cases['mean']=np.mean(c0_local_sumcount_changes_300_cases,axis=1)\n",
    "    c0_local_sumcount_changes_300_controls['mean']=np.mean(c0_local_sumcount_changes_300_controls,axis=1)\n",
    "    ######sumprob\n",
    "    c0_local_sumprob_ratios_odd100_cases['mean']=np.mean(c0_local_sumprob_ratios_odd100_cases,axis=1)\n",
    "    c0_local_sumprob_ratios_odd100_controls['mean']=np.mean(c0_local_sumprob_ratios_odd100_controls,axis=1)\n",
    "    c0_local_sumprob_ratios_odd200_cases['mean']=np.mean(c0_local_sumprob_ratios_odd200_cases,axis=1)\n",
    "    c0_local_sumprob_ratios_odd200_controls['mean']=np.mean(c0_local_sumprob_ratios_odd200_controls,axis=1)\n",
    "    c0_local_sumprob_ratios_odd300_cases['mean']=np.mean(c0_local_sumprob_ratios_odd300_cases,axis=1)\n",
    "    c0_local_sumprob_ratios_odd300_controls['mean']=np.mean(c0_local_sumprob_ratios_odd300_controls,axis=1)\n",
    "    #####\n",
    "    dict_var_Score_sumprob_cases_100_odds[name+'_cases_logsumprob']=c0_local_sumprob_ratios_odd100_cases\n",
    "    dict_var_Score_sumprob_controls_100_odds[name+'_controls_logsumprob']=c0_local_sumprob_ratios_odd100_controls\n",
    "    #\n",
    "    dict_var_Score_sumcounts_cases_100[name+'_cases_logsumcounts']=c0_local_sumcount_changes_100_cases\n",
    "    dict_var_Score_sumcounts_controls_100[name+'_cases_logsumcounts']=c0_local_sumcount_changes_100_controls\n",
    "\n",
    "\n",
    "\n",
    "    \n",
    "    \n",
    "    \n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
